### PR TITLE
feat(samples): Full Kubeflow E2E sample using KFPv2.

### PR DESCRIPTION
**Pull request by zijianjoy**
**Description of your changes:**

Validate the KFPv2 syntax is runnable which connects multiple Kubeflow components: (Katib, TFJob, KServe, MLMD)

We make use of multiple KFPv2 features in this example:
- Make use of multiple primitive parameters.
- Make use of Input/Output Artifact.
- Make use of lightweight python component.
- Make use of custom container component.
- Make use of pipeline-in-pipeline.
- Make use of `after()` dependent tasks.

Issues discovered:
- https://github.com/kubeflow/pipelines/issues/8300
- https://github.com/kubeflow/pipelines/issues/8302
- Question: Does container component accept any parameter type other than string?
- FE: Sub-DAG visualization is not complete. This is expected.

pipeline_run_detail_mnist_v2

/assign @chensun @connor-mccarthy 

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Review by connor-mccarthy — CHANGES_REQUESTED** [review]
This is a great sample, @zijianjoy. I'm not familiar enough with katib and kserve to make any content suggestions for those aspects, but I do have one comment about component output annotations.

Thanks for this!

File: samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb [hunk 1]
```diff
@@ -0,0 +1,782 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# You should define the Experiment name, namespace and number of training steps in the arguments.\n",
+    "@dsl.component(\n",
+    "    base_image='python:3.8',\n",
+    "    packages_to_install=['kubeflow-katib==0.12.0']\n",
+    ")\n",
+    "def create_katib_experiment_task(experiment_name: str, experiment_namespace: str, training_steps: str,\n",
+    "                                experiment_spec_json: OutputPath(Dict[str, str])):\n",
```
[the remark this answers is not in the document]
Comment (zijianjoy) [reply samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb]: Thank you Connor for the insight! I have fixed the usage for a preferred approach of output.

File: samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb [hunk 2]
```diff
@@ -0,0 +1,777 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# You should define the Experiment name, namespace and number of training steps in the arguments.\n",
+    "@dsl.component(\n",
+    "    base_image='python:3.8',\n",
+    "    packages_to_install=['kubeflow-katib==0.12.0']\n",
+    ")\n",
+    "def create_katib_experiment_task(experiment_name: str, experiment_namespace: str, training_steps: str\n",
+    "                                ) -> NamedTuple('Outputs', [('experiment_spec_json', Dict[str, str])]):\n",
```
Comment (connor-mccarthy) [inline samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb:149]: `NamedTuple` with one output should probably just be a single output (`-> Dict[str, str]`)

This simplifies the component data passing logic when constructing the pipeline from `task.outputs['experiment_spec_json']` to `task.output`.

File: samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb [hunk 3]
```diff
@@ -0,0 +1,777 @@
...
+    "        }\n",
+    "    }\n",
+    "    \n",
+    "    # Convert chief spec to String type Output.\n",
+    "    import json\n",
+    "    chief_spec_str = json.dumps(tfjob_chief_spec)\n",
+    "    \n",
+    "    # Convert worker spec to String type Output.\n",
+    "    import json\n",
+    "    worker_spec_str = json.dumps(tfjob_worker_spec)\n",
```
Comment (connor-mccarthy) [inline samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb:466]: Do you think it makes sense to return this (and `tfjob_chief_spec`) as a dict, letting the lightweight component backend handle the serialization/deserialization?

File: samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb [hunk 4]
```diff
@@ -0,0 +1,777 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@dsl.component\n",
+    "def create_serving_task(model_name: str, model_namespace: str, model_volume_name: str\n",
+    "                       ) -> NamedTuple('Outputs', [('inferenceservice_yaml', Dict[str, str])]):\n",
```
Comment (connor-mccarthy) [inline samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb:520]: ditto: `NamedTuple` with one output

File: samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb [hunk 5]
```diff
@@ -0,0 +1,777 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "name=\"mnist-e2e\"\n",
+    "namespace=\"kubeflow-user-project\"\n",
+    "training_steps=\"200\"\n",
+    "model_volume_name=\"workflow1-model-volume\"\n",
+    "\n",
+    "\n",
+    "import json\n",
```
Comment (connor-mccarthy) [inline samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb:620]: nit: `json` not used after this cell

**Review by connor-mccarthy — CHANGES_REQUESTED** [review]
Left a few comments, all small things.

Just to confirm, have you validated that this pipeline runs?

File: samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb (hunk 2 above)
Comment (connor-mccarthy) [reply samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb:149]: Discussed offline. Prefer the named output for this example.

File: samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb (hunk 4 above)
Comment (connor-mccarthy) [reply samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb:520]: https://github.com/kubeflow/pipelines/pull/8303#discussion_r980591550

File: samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb (hunk 3 above)
Comment (zijianjoy) [reply samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb:465]: Thank you for the comment! I like the idea of using dict. However, These two values are needed in the container component below: `tfjob_launcher`. If I understand it correctly, I have to convert Dict to str for container component input. Correct?

File: samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb (hunk 5 above)
Comment (zijianjoy) [reply samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb]: Good point! Cleaned up unnecessary `import json`

File: samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb (hunk 3 above)
Comment (connor-mccarthy) [reply samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb:465]: You can pass the dict directly to the container component. The conversion to a string happens implicitly when the commands/args are provided at container runtime.

As a minimal example, the following is a valid pipeline:

```python
from kfp import dsl
from typing import Dict


@dsl.component
def comp1() -> Dict[str, str]:
    return {'string': 'string'}


@dsl.container_component
def container(dictionary: Dict[str, str]):
    return dsl.ContainerSpec(
        image='alpine',
        command=['echo', dictionary],
    )


@dsl.pipeline
def my_pipeline():
    op1 = comp1()
    op2 = container(dictionary=op1.output)
```

Comment (zijianjoy) [reply samples/contrib/kubeflow-e2e-mnist/kubeflow-e2e-mnist-kfpv2.ipynb]: It is nice to know! Thank you Connor and I am able to make use of the said feature to convert Dict to str.

**Comment by connor-mccarthy** [comment]
This PR has identified a bug in KFP SDK 2.0.0b5.

Let's wait to merge until this is fixed by https://github.com/kubeflow/pipelines/pull/8316.

**Comment by connor-mccarthy** [comment]
#8316 is merged, but not yet released.

/lgtm

**Comment by connor-mccarthy** [comment]
^ this is an error -- #8444 doesn't fix this

**Pull request merged**